core/src/main/java/org/infinispan/interceptors/impl/BaseStateTransferInterceptor.java

```diff
@@ -32,53 +46,37 @@
    private final boolean trace = getLog().isTraceEnabled();
 
    protected StateTransferManager stateTransferManager;
-   private StateTransferLock stateTransferLock;
+   protected StateTransferLock stateTransferLock;
+   protected Executor remoteExecutor;
    private GroupManager groupManager;
    private long transactionDataTimeout;
+   private ScheduledExecutorService timeoutExecutor;
 
    @Inject
    public void init(StateTransferLock stateTransferLock, Configuration configuration,
-                    StateTransferManager stateTransferManager, GroupManager groupManager) {
+                    StateTransferManager stateTransferManager, GroupManager groupManager,
+                    @ComponentName(KnownComponentNames.TIMEOUT_SCHEDULE_EXECUTOR) ScheduledExecutorService timeoutExecutor,
+                    @ComponentName(KnownComponentNames.REMOTE_COMMAND_EXECUTOR) Executor remoteExecutor) {
       this.stateTransferLock = stateTransferLock;
       this.stateTransferManager = stateTransferManager;
       this.groupManager = groupManager;
+      this.timeoutExecutor = timeoutExecutor;
+      this.remoteExecutor = remoteExecutor;
       transactionDataTimeout = configuration.clustering().remoteTimeout();
    }
 
    @Override
    public BasicInvocationStage visitGetKeysInGroupCommand(InvocationContext ctx, GetKeysInGroupCommand command) throws Throwable {
-      final String groupName = command.getGroupName();
-      final boolean isOwner = groupManager.isOwner(groupName);
       updateTopologyId(command);
-      final int commandTopologyId = command.getTopologyId();
 
       if (ctx.isOriginLocal()) {
-         return invokeNext(ctx, command).compose((stage, rCtx, rCommand, rv, t) -> {
-            boolean shouldRetry;
-            if (t != null) {
-               // Must retry if we got an OutdatedTopologyException or SuspectException
-               Throwable ce = t;
-               while (ce instanceof RemoteException) {
-                  ce = ce.getCause();
-               }
-               shouldRetry = ce instanceof OutdatedTopologyException || ce instanceof SuspectException;
-            } else {
-               // Only check the topology id if if we are an owner
-               shouldRetry = isOwner && currentTopologyId() != commandTopologyId;
-            }
-            if (shouldRetry) {
-               return retryVisitGetKeysInGroupCommand(ctx, command, commandTopologyId);
-            }
-            // No retry, either rethrow the exception or return the current result
-            if (t != null) {
-               throw t;
-            } else {
-               return returnWith(rv);
-            }
-         });
+         return invokeNext(ctx, command).compose(this::handleGeyKeysInGroupReturn);
```

Comment:
Yeah, looking at it again it may make things more complicated. How about renaming it `handleLocalGeyKeysInGroupReturn` instead?